File: Programers/괄호변환.java
import java.util.*;
class Solution {
    int position;
    public String solution(String p) {
        
        if(p.isEmpty()) return p;
        
        boolean correct = isCorrect(p);
        
        String u = p.substring(0,position);
        String v = p.substring(position,p.length());
        
        if(correct){
            return u + solution(v);
        }
        
        String answer = "(" + solution(v) + ")";
        
        for(int i=1; i < u.length()-1; i++){
            if(u.charAt(i)== '('){
                answer += ")";
            }else{
                answer += "(";
            }
        }
        
        return answer;
    }
    boolean isCorrect(String str){
        boolean ret = true;
        int left = 0, right= 0;
        Stack<Character>st = new Stack<>();
        
        for(int i= 0 ; i < str.length(); i++){
            if(str.charAt(i) == '('){
                left++;
                st.push('(');
            }
            else{
                right++;
                if(st.isEmpty()){
                    ret = false;
                }else{
                    st.pop();
                }
            }
            if(left == right){
                position = i+1;
                return ret;
            }
        }
        return true;
    }
}

File: Programers/괄호변환.js
function solution(p) {
    let answer = '';
    let open = 0;
    let close = 0;
    
    if(!p) return p;
    
    for(let i=0; i< p.length; i++){
        if(p[i] === "(") open++;
        else close++
        
        if(open === close){
            if(check(p.slice(0,i+1))){
                answer  = p.slice(0,i+1) + solution(p.slice(i+1));
                return answer;
            }else {
                answer = '(' + solution(p.slice(i+1)) + ')';
                
                for (let j=1; j<i; j++) {
                    if (p[j] === '(') {
                        answer = answer + ')'
                    } else {
                        answer = answer + '('
                    }
                }
                return answer;
            }
        }
    }
    
    return answer;
}

function check(p){
    let stack = [];
    for(let i=0; i< p.length; i++){
        if(p[i] === "(") stack.push("(")
        else{
            if(stack.length === 0) return false;
            stack.pop();
        }
    }
    return true;
}
